File: frontend/shader/contract/ts/tron.ts
// ============================================================================
// Tron Configuration
// ============================================================================

export interface TronConfig {
  rpcUrl: string;
  network: "mainnet" | "shasta" | "nile";
}

// ============================================================================
// Tron API Response Types
// ============================================================================

export interface TronAccountResponse {
  balance?: string;
}

/**
 * Tron Connector
 */
export class TronConnector {
  private readonly config: TronConfig;

  constructor(config: TronConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Tron address");
    }

    try {
      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = "https://api.trongrid.io";
      } else if (this.config.network === "shasta") {
        apiUrl = "https://api.shasta.trongrid.io";
      } else {
        apiUrl = this.config.rpcUrl;
      }

      const response = await fetch(`${apiUrl}/v1/accounts/${address}`, {
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        throw new Error("Failed to fetch Tron balance");
      }

      const data = (await response.json()) as TronAccountResponse;
      // Return balance in sun (smallest unit, 1 TRX = 1,000,000 sun)
      return data.balance?.toString() || "0";
    } catch (error) {
      throw new Error(
        `Tron getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Tron address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Tron transactions require:
      // 1. Creating transaction
      // 2. Signing with private key
      // 3. Broadcasting to network
      // This is simplified - in production use tronweb

      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = "https://api.trongrid.io";
      } else if (this.config.network === "shasta") {
        apiUrl = "https://api.shasta.trongrid.io";
      } else {
        apiUrl = this.config.rpcUrl;
      }

      // Create transaction
      const createResponse = await fetch(
        `${apiUrl}/wallet/createtransaction`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            to_address: toAddress,
            owner_address: fromAddress,
            amount: amountNum,
          }),
        }
      );

      if (!createResponse.ok) {
        throw new Error("Failed to create Tron transaction");
      }

      const transaction = await createResponse.json();

      // Sign and broadcast
      // Note: In production, sign with private key before broadcasting
      const broadcastResponse = await fetch(
        `${apiUrl}/wallet/broadcasttransaction`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(transaction),
        }
      );

      if (!broadcastResponse.ok) {
        throw new Error("Failed to broadcast Tron transaction");
      }

      const broadcastData = (await broadcastResponse.json()) as {
        txid?: string;
      };
      return (
        broadcastData.txid || (transaction as any).txID || "transaction_id"
      );
    } catch (error) {
      throw new Error(
        `Tron sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    return address.length === 34 && address.startsWith("T");
  }

  /**
   * Bridge tokens from Tron to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Tron";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["tron", "trx"];
  }
}

File: store/rust/rust/tron.rs
//! Tron connector

use crate::{BridgeError, BridgeResult, TronNetwork, http_post_json};
use serde::{Deserialize, Serialize};

/// Tron connector
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TronConnector {
    rpc_url: String,
    #[allow(dead_code)]
    network: TronNetwork,
}

impl TronConnector {
    pub fn new(rpc_url: &str, network: TronNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Tron API call
        let url = format!("{}/wallet/getaccount", self.rpc_url.trim_end_matches('/'));
        let body = serde_json::json!({
            "address": address,
            "visible": true
        });

        let result = http_post_json(&url, body)?;
        let balance = result
            .get("balance")
            .and_then(|b| b.as_u64())
            .or_else(|| {
                result
                    .get("balance")
                    .and_then(|b| b.as_str())
                    .and_then(|s| s.parse::<u64>().ok())
            })
            .unwrap_or(0);

        Ok(balance)
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in sun (1 TRX = 1,000,000 sun)
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid Tron address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Tron transaction creation and broadcast
        let url = format!("{}/wallet/createtransaction", self.rpc_url.trim_end_matches('/'));
        let body = serde_json::json!({
            "to_address": to_address,
            "owner_address": from_address,
            "amount": amount
        });

        let result = http_post_json(&url, body)?;
        result
            .get("txID")
            .or_else(|| result.get("txid"))
            .and_then(|t| t.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| BridgeError::TransactionFailed("Failed to create Tron transaction".to_string()))
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // Tron addresses are base58, 34 chars, start with 'T'
        address.len() == 34 && address.starts_with('T')
    }
}

